### PR TITLE
Scrum 128/handlers lib for error and operation handling

### PR DESCRIPTION
**Pull request by vsh51**
Only Handlers should be reviewed. SMTPresponse is being reviewed on another pull request

### CODE REVIEW

File: libs/Handlers/include/Handlers.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+#include <boost/system/error_code.hpp>
+#include <boost/asio.hpp>
+#include <iostream>
+
+#include "SMTPResponse.h"
+
+using std::string;
+
+namespace ISXLOGS
```
Comment (zikkend) [inline libs/Handlers/include/Handlers.h]: Rename ISXLOGS to ISXLogs. According to code convention namespaces should be in following format: ISXCamelCase

File: libs/Handlers/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (zikkend) [inline libs/Handlers/CMakeLists.txt]: Maybe should use newer version of cmake

File: libs/Handlers/include/Handlers.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+{
+class SmartSocketMethodsHandlers
+{
+public:
+    static bool HandleConnection(
+        const string& server, const int port
```
Comment (zikkend) [inline libs/Handlers/include/Handlers.h:17]: Consider use of string_view intead of string&

File: libs/Handlers/src/Handlers.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        return true;
+    }
+
+    HandleError("Close error", error_code);
+    return false;
+};
+
+bool SmartSocketMethodsHandlers::HandleUpgradeSecurity(
+        const boost::system::error_code& error_code
+        , bool* ssl_toogle)
```
Comment (zikkend) [inline libs/Handlers/src/Handlers.cpp]: Maybe misspelled toogle with toggle

File: libs/SMTPResponse/src/SMTPResponse.cpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    bool SMTPResponse::IsValidResponse(const std::string& response, std::smatch& matches) const
+    {
+        std::regex responsePattern(R"(^(\d{3})(?:[ -](\d\.\d\.\d))?[ -](.*)$)");
+        return std::regex_match(response, matches, responsePattern);
+    }
+
+    auto SMTPResponse::SplitAtEndline(const std::string& response) const -> std::pair<std::string, std::string>
+    {
+        auto endline_pos = std::min(response.find("\n"), response.find("\r"));
```
Comment (zikkend) [inline libs/SMTPResponse/src/SMTPResponse.cpp:109]: Maybe, there is need to check endline_pos for npos

**Review by zikkend — APPROVED** [review]
Overall, code is well written and well structured

File: libs/Handlers/include/Handlers.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+private:
+    static inline void HandleError(
+        const string& prefix, const boost::system::error_code& error_code);
+        
+    SmartSocketMethodsHandlers() = delete;
+    ~SmartSocketMethodsHandlers() = delete;
+
+    static inline std::ostream* s_log_stream = &std::clog;
+};
+}; // namespace ISXErrorHandlers
```
Comment (TheCheezyPie) [inline libs/Handlers/include/Handlers.h]: Am I tweaking or this doesn't match the namespace declaration on line 11?

File: libs/Handlers/src/Handlers.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        *s_log_stream << "LOG: " << "Handshake successful. Connection upgraded to TLS" << std::endl;
+        *ssl_toogle = true;
+        return true;
+    }
+
+    HandleError("Update security error", error_code);
+    *ssl_toogle = false;
+    return false;
+};
+}; // namespace ISXErrorHandlers
```
Comment (TheCheezyPie) [inline libs/Handlers/src/Handlers.cpp]: doesn't match the line 5 namespace name (or i'm seeing things)

File: libs/SMTPResponse/include/SMTPResponse.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        std::string get_enhanced_code() const;
+        std::string get_text() const;
+    
+        StatusType get_status() const;
+        bool CodeEquals(u_int16_t code) const;
+        bool StatusEquals(StatusType status) const;
+        std::string get_formated_response() const;
+    
+    private:
+        u_int16_t m_code;
```
Comment (TheCheezyPie) [inline libs/SMTPResponse/include/SMTPResponse.h]: Data members should be the lowest

File: libs/SMTPResponse/include/SMTPResponse.h [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+#include "StatusType.h"
+
+namespace ISXR
+{
+
+    class SMTPResponse {
+    public:
+        explicit SMTPResponse(const std::string& response);
+    
+        u_int16_t get_code() const;
```
Comment (TheCheezyPie) [inline libs/SMTPResponse/include/SMTPResponse.h]: consider making these inline

File: libs/SMTPResponse/src/SMTPResponse.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        return m_code == code;
+    }
+
+    bool SMTPResponse::StatusEquals(StatusType status) const 
+    {
+        return m_status == status;
+    }
+
+} // namespace SMTPResonse
```
Comment (TheCheezyPie) [inline libs/SMTPResponse/src/SMTPResponse.cpp]: isn't it namespace ISXR? and also a typo. SMTPRes**P**onse

**Review by TheCheezyPie — APPROVED** [review]
Once again, I am late to the party, the code looks good with very little room for improvement

File: libs/SMTPResponse/src/SMTPResponse.cpp (hunk 5 above)
Comment (vsh51) [reply libs/SMTPResponse/src/SMTPResponse.cpp:109]: Under review on another branch

File: libs/SMTPResponse/include/SMTPResponse.h (hunk 8 above)
Comment (vsh51) [reply libs/SMTPResponse/include/SMTPResponse.h]: Under review on another branch

File: libs/SMTPResponse/include/SMTPResponse.h (hunk 9 above)
Comment (vsh51) [reply libs/SMTPResponse/include/SMTPResponse.h]: Under review on another branch

File: libs/SMTPResponse/src/SMTPResponse.cpp (hunk 10 above)
Comment (vsh51) [reply libs/SMTPResponse/src/SMTPResponse.cpp]: Under review on another branch

File: libs/SMTPResponse/src/SMTPResponse.cpp (hunk 5 above)
Comment (RomaB-5) [reply libs/SMTPResponse/src/SMTPResponse.cpp:109]: > Maybe, there is need to check endline_pos for npos

Not really. The value endline_pos is used later only in two places.
```cpp
response.substr(0, endline_pos)
``` 
- where second parameter is default to std::string::npos
```cpp
response.substr(std::min(endline_pos, response.size())))
```
- and here where there is a clamp.

There is no real need to complicate a simple function with extra checks.

**Pull request closed without merging**